zeebe/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/RecordCounterTest.java

```diff
@@ -0,0 +1,231 @@
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.camunda.zeebe.exporter.test.ExporterTestConfiguration;
+import io.camunda.zeebe.exporter.test.ExporterTestContext;
+import io.camunda.zeebe.exporter.test.ExporterTestController;
+import io.camunda.zeebe.protocol.record.Record;
+import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.test.util.junit.RegressionTest;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class RecordCounterTest {
+  private final ElasticsearchExporterConfiguration config =
+      new ElasticsearchExporterConfiguration();
+  private final ExporterTestContext context =
+      new ExporterTestContext().setConfiguration(new ExporterTestConfiguration<>("test", config));
+  private final ExporterTestController controller = new ExporterTestController();
+  private final ElasticsearchClient client = mock(ElasticsearchClient.class);
+  private final ElasticsearchExporter exporter =
+      new ElasticsearchExporter() {
+        @Override
+        protected ElasticsearchClient createClient() {
+          return client;
+        }
+      };
+
+  @BeforeEach
+  void beforeEach() {
+    when(client.putIndexTemplate(any())).thenReturn(true);
+    when(client.putComponentTemplate()).thenReturn(true);
+  }
+
+  @Test
+  void shouldIncrementCounterOnSynchronousFlush() {
+    // given
+    exporter.configure(context);
+    exporter.open(controller);
+    final Record mockRecord = mock(Record.class);
+    final var valueType = ValueType.PROCESS_INSTANCE;
+    when(mockRecord.getValueType()).thenReturn(valueType);
+
+    // when a new record is exported
+    when(client.shouldFlush()).thenReturn(true);
+    exporter.export(mockRecord);
+
+    // then the record counter should be 1
+    final var counters = readMetadata();
+    assertThat(counters).describedAs("The counter is stored in the metadata").isNotEmpty();
+    assertThat(counters.get(valueType))
+        .describedAs("The record counter should be 1 as only exported 1 record")
+        .isEqualTo(1);
+  }
+
+  @Test
+  void shouldNotIncrementCounterOnFailedSynchronousFlush() {
+    // given
+    exporter.configure(context);
+    exporter.open(controller);
+    final Record mockRecord = mock(Record.class);
+    final var valueType = ValueType.PROCESS_INSTANCE;
+    when(mockRecord.getValueType()).thenReturn(valueType);
+
+    // when a new record is exported, but the flush fails
+    when(client.shouldFlush()).thenReturn(true);
+    doThrow(new ElasticsearchExporterException("failed to flush")).when(client).flush();
+    assertThatThrownBy(() -> exporter.export(mockRecord))
+        .isInstanceOf(ElasticsearchExporterException.class);
+
+    // then the record counter should be empty
+    final var counters = readMetadata();
+    assertThat(counters).describedAs("The counter is not stored in the metadata").isEmpty();
```

Comment:
@nicpuppa But isn't the whole point that the counter should be incremented regardless of the flush succeeding or failing?

Reply:
The in-memory counter is updated (when zeebe index the record in the batch), but if the flush fails we don't save the metadata in the state. This test basically verifies this behaviour.

Reply:
@remcowesterhoud @nicpuppa okay, but then we still miss a test case for the actual bug right?

Reply:
[This](https://github.com/camunda/camunda/pull/24214/files#diff-776942713372ad3539356e7bd710c6c27c1ee13c7da16c1f41103b3f535c7cadR142-R170) test covers the actual bug. Here we fail on the first flush and perform an asynchronous flush. The counter in the controller is 1. This counter is takes from the internal counter on a successful flush, so it's safe to assume the internal counter is 1 as well. Before the fix this would be 0 and the next record would reuse 1.

Reply:
Oh I see what you mean we could add something like:

```java
    @Test
     void shouldIncrementCounterOnFlushErrors() {
       // given
       when(client.shouldFlush()).thenReturn(true);

       final var record = newRecord(PARTITION_ID, ValueType.PROCESS_INSTANCE);

       // when
       doThrow(new ElasticsearchExporterException("failed to flush")).when(client).flush();

       assertThatCode(() -> exporter.export(record))
           .isInstanceOf(ElasticsearchExporterException.class);

       // then
      final var recordSequenceCaptor = ArgumentCaptor.forClass(RecordSequence.class);
      verify(client, times(1)).index(any(), recordSequenceCaptor.capture());
      assertThat(recordSequenceCaptor.getAllValues())
          .extracting(RecordSequence::counter)
          .describedAs("Expect that the record counter is the same on retry")
          .containsExactly(1L);
    }
```

Reply:
Yes! Thanks!